src/torch/README.md

```diff
@@ -0,0 +1,131 @@
+* `torch.autograd.Function`: the `forward` method (forward pass) is the function evaluation. The `backward` method (backward pass) computes the VJP. More specifically, the `backward(*grad_output)` method multiplies the `grad_output` which represents the gradient(s) of a subsequent scalar-valued objective function with respect to the output of `forward`, via chain-rule, by the adjoint of the Jacobian of the `forward` method. 
+
+This SIRF-PyTorch wrapper is **only** for reverse-mode automatic differentiation via subclassing `torch.autograd.Function`.
+
+## Wrapper Design
+
+The wrapper provides three main classes:
+
+1.  `SIRFTorchOperator`: Wraps a SIRF `Operator` (e.g., a projection operator). Applies the operator forward pass, and applies the adjoint in backward pass.
+2.  `SIRFTorchObjectiveFunction`: Wraps a SIRF `ObjectiveFunction` for computes the value in the forward pass, and objective function gradient in backward pass.
+3.  `SIRFTorchObjectiveFunctionGradient`: Wraps a SIRF `ObjectiveFunction` that computes the objective function gradients in the forward pass and the Hessian-vector product in the backward pass. In the backward the Hessian is evaluated at the point which the objective functions gradient was evaluated.
+
+These classes use custom `torch.autograd.Function` implementations (`_Operator`, `_ObjectiveFunction`, and `_ObjectiveFunctionGradient`) to define the forward and backward passes, handling the conversions between PyTorch tensors and SIRF objects.
+
+### `_Operator` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Applies the SIRF `Operator.forward()` method.
+    3.  Converts the result back to a PyTorch tensor.
+    4.  If the input tensor requires gradients, it saves relevant information (the output SIRF object and the operator) in the context (`ctx`) for use in the backward pass.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the "upstream gradient" (`grad_output`).
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Applies the SIRF `Operator.backward()` method. This will apply the **Jacobian adjoint** of the operator to upstream gradient (the vector).
+    4.  Converts the resulting SIRF object back to a PyTorch tensor and returns it.
+
+### `_ObjectiveFunction` (Forward and Backward Passes)
+
+*   **Forward Pass:**
+    1.  Converts the input PyTorch tensor (representing an image) to a SIRF object.
+    2.  Calls the SIRF `ObjectiveFunction.get_value()` method.
+    3.  Returns the of the objective function value as a PyTorch tensor.
+    4. Saves relevant information to the `ctx` if gradients are needed.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), in this case it is always a scalar.
+    2.  Gets the gradient of the objective function using `sirf_obj_func.get_gradient()`, which computed at the input and multiplied by the upstream gradient.
```

Comment:
@paskino, what does CIL use? `gradient` or `get_gradient`? I'm guessing SIRF uses both, so we should probably use the CIL convention here, such that the wrapper is CIL compatible.

Reply:
I use `gradient` in the code will update this text